stdlib/gc.mli

```diff
@@ -491,39 +495,70 @@ module Memprof :
        recorded at every sample. Its default is [max_int].
 
        The parameter [tracker] determines how to track sampled blocks
        over their lifetime in the minor and major heap.
 
-       Sampling is temporarily disabled when calling a callback
-       for the current thread. So they do not need to be re-entrant if
-       the program is single-threaded. However, if threads are used,
-       it is possible that a context switch occurs during a callback,
-       in this case the callback functions must be re-entrant.
+       Sampling is temporarily disabled on the current thread when
+       calling a callback, so callbacks do not need to be re-entrant
+       if the program is single-threaded and single-domain. However,
+       if threads or multiple domains are used, it is possible that
+       several callbacks will run in parallel. In this case, callback
+       functions must be re-entrant.
+
+       Note that a callback may be postponed slightly after the actual
+       event. The callstack passed to an allocation callback always
+       accurately reflects the allocation, but the program state may
+       have evolved between the allocation and the call to the
+       callback.
+
+       If a new thread or domain is created when profiling is active,
+       the child thread or domain joins that profile (using the same
+       [sampling_rate], [callstack_size], and [tracker] callbacks).
 
-       Note that the callback can be postponed slightly after the
-       actual event. The callstack passed to the callback is always
-       accurate, but the program state may have evolved. *)
+       An allocation callback is generally run by the thread which
+       allocated the block. If the thread exits or the profile is
+       stopped before the callback is called, the callback may be run
+       by a different thread.
+
+       Each callback is generally run by the domain which allocated
+       the block. If the domain terminates or the profile is stopped
+       before the callback is called, the callback may be run by a
+       different domain.
+
+       Different domains may run different profiles simultaneously.
+       *)
 
     val stop : unit -> unit
-    (** Stop the sampling. Fails if sampling is not active.
 
-        This function does not allocate memory.
+    (** Stop sampling for the current profile. Fails if no profile is
+       sampling in the current domain. Stops sampling in all threads
+       and domains sharing the profile.
+
+       Callbacks from a profile may run after [stop] is called, until
+       [discard] is called on the return value of [start].
+
+       A profile is implicitly stopped (but not discarded) if all
+       domains and threads sampling for it are terminated.
+       *)
 
-        All the already tracked blocks are discarded. If there are
-        pending postponed callbacks, they may be discarded.
+    val discard : t -> unit
```

Comment:
This API is somewhat incoherent, in that it requires a profile value while all the other functions use the "current profile". Any reason for this? Why would discard not work with the current profile?

Reply:
I have to speculate on how this API is implemented, but reasoning in terms of ownership I would expect `stop` to transfer ownership of the callback state, rather than `start`; i.e.
```ocaml
val start : [...] -> unit
val stop : [...] -> t
```
This would address your point. Again this is making assumptions on what is natural from an implementation standpoint but we don't have the implementation.

Reply:
Clearly more documentation and/or a better interface is required here. The intention is to allow uses such as a sequence of `start`/`stop` pairs, followed by subsequent `discard` operations on some or all of the resulting profiles. "The current profile" is only meaningful between `start` and `stop`.

For example, some highly-phased system (such as a traditional compiler) which is interested in the long-term fate of objects allocated during some particular phase or phases (say: objects allocated during some specific compiler optimization phase). It will call `start` and `stop` around the phase, but may not call `discard` until much later, giving time for allocated objects to be promoted and eventually GCed.

In such a use case, `discard` is still necessary to give the profiling tool author more control (before `discard`, as callbacks may run, that profile may cause overhead; after `discard`, it will not).

The primary motivation of this change is to allow callbacks to run after sampling has stopped. The rest of this design follows from that. (I agree with @gadmm that `stop` could return the `t`. Would that be better?)

Speculation isn't really necessary; #12379 (although still messy and buggy) fleshes this interface out. `t` is a disguised integer, the serial number of an underlying C structure which links to domain- and thread-specific data structures (and inherits objects from them if they terminate).

Reply:
I had looked at #12379 but I find it hard to read at the moment, and having an overview is helpful anyway.

I have more questions and possible suggestions but I'd normally look at the implementation first. Can you please pinpoint the relevant commits implementing this change? (either now or after you finish your cleanup of #12379, there is no rush.) (I am thinking about another strategy to fully decouple stopping sampling and stopping running callbacks: implementing `t` as an atomic flag stored on the OCaml heap. The callbacks are not actually removed from the queues but simply ignored if the flag is set. Then it even makes sense to call `discard` before calling `stop`. But I might be missing some details at the moment.)